Generate serialization for CNContact

beidson opened the pull request:
#### 5ae2c85df7cb2b23c012f3d0c684f29ff660face
<pre>
Generate serialization for CNContact
<a href="https://bugs.webkit.org/show_bug.cgi?id=269153">https://bugs.webkit.org/show_bug.cgi?id=269153</a>

Reviewed by Alex Christensen.

* Source/WebCore/PAL/pal/cocoa/ContactsSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/ContactsSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/ContactsSPI.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.h:
(WebKit::CoreIPCCNContact::CoreIPCCNContact):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.mm:
(WebKit::CoreIPCCNContact::CoreIPCCNContact):
(WebKit::CoreIPCCNContact::isValidCNContactType):
(WebKit::CoreIPCCNContact::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/274525@main">https://commits.webkit.org/274525@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/845ce7208dd43ff6f805981bed065311e15c4457

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39355 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41708 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41889 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21199 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15663 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39929 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34086 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13406 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43167 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35719 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35356 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39173 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/11676 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37417 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15773 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8798 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCContacts.h:
```diff
@@ -89,6 +90,49 @@ class CoreIPCCNPhoneNumber {
+    String m_identifier;
+    NSInteger m_contactType;
+
+    std::optional<String> m_namePrefix;
```
I don't think these need to be optional.  String already has a null state.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCContacts.mm:
```diff
@@ -74,6 +74,170 @@
+    result.get().contactType = (CNContactType)m_contactType;
+
+    if (m_namePrefix)
+        result.get().namePrefix = *m_namePrefix;
```
if (!m_namePrefix.isNull())
    result.get().namePrefix = m_namePrefix;

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCContacts.h:
```diff
@@ -89,6 +90,49 @@ class CoreIPCCNPhoneNumber {
+    std::optional<CoreIPCDateComponents> m_birthday;
+    std::optional<CoreIPCDateComponents> m_nonGregorianBirthday;
+
+    Vector<std::tuple<String, String, CoreIPCDateComponents>> m_dates;
```
I don't think we should use tuple.  Give the pieces a name with a nested struct.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCContacts.mm:
```diff
@@ -74,6 +74,170 @@
+        for (auto& tuple : m_urlAddresses) {
+            RetainPtr<CNLabeledValue> labeledValue = adoptNS([[PAL::getCNLabeledValueClass() alloc] initWithIdentifier:std::get<0>(tuple) label:std::get<1>(tuple) value:std::get<2>(tuple)]);
+            [labeledValues addObject:labeledValue.get()];
+        }
```
createNSArray can be used to greatly simplify all of these to a one line lambda.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCContacts.h:
```diff
@@ -89,6 +91,63 @@ class CoreIPCCNPhoneNumber {
+
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCCNContact, void>;
+    CoreIPCCNContact()
```
= default

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCContacts.h:
```diff
@@ -89,6 +91,63 @@ class CoreIPCCNPhoneNumber {
+    }
+
+    String m_identifier;
+    NSInteger m_contactType;
```
{ 0 }

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1211,6 +1211,8 @@ - (instancetype)initWithMerchantIdentifier:(NSString *)merchantIdentifier
     RetainPtr<CNPostalAddress> address = postalAddressForTesting();
     RetainPtr<CNLabeledValue> labeledPostalAddress = adoptNS([[PAL::getCNLabeledValueClass() alloc] initWithLabel:@"Work" value:address.get()]);
 
+    RetainPtr<CNLabeledValue> labeledEmailAddress = adoptNS([[PAL::getCNLabeledValueClass() alloc] initWithLabel:@"WorkSPAM" value:@"spam@webkit.org"]);
```
auto

beidson committed:
Generate serialization for CNContact
https://bugs.webkit.org/show_bug.cgi?id=269153

Reviewed by Alex Christensen.

* Source/WebCore/PAL/pal/cocoa/ContactsSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/ContactsSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/ContactsSPI.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSObject<NSSecureCoding>>):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.h:
(WebKit::CoreIPCCNContact::CoreIPCCNContact):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.mm:
(WebKit::CoreIPCCNContact::CoreIPCCNContact):
(WebKit::CoreIPCCNContact::isValidCNContactType):
(WebKit::CoreIPCCNContact::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: https://commits.webkit.org/274525@main
